### ISSUE TITLE
Issue with StepGetVNCPort

### ISSUE THREAD

**Issue by chronidev** [xenserver/packer-builder-xenserver#47]
There is a  problem with [StepGetVNCPort](https://github.com/xenserver/packer-builder-xenserver/blob/master/builder/xenserver/common/step_get_vnc_port.go#L12-L24), when you try to use  this plugin on a xen pool.

The problem is that you try to retrieve vnc port through the xenstore using ssh, but on a pool the vm could be created on a different xenserver than one used for the initial connection, and the xenstore is local to one xenserver not to a pool, so packer failed.

Maybe a solution should be to use the API instead of using xenstore through ssh, or maybe iterate over all members of the xen pool to find where the vm was created.

**Comment by chronidev** [comment]
I read the API documentation and it's not possible to retrieve the vnc port through it. So the solution I found it's to retrieve the location(xenserver host) of the console, then I connect to this xenserver then I read the xenstore.  I have a dirty patch (it's first I touch golang), I will try to clean that mess before send a PR.

**Comment by darnells** [comment]
@chronidev I am having the same problem. Would you share your patch in spite of its dirtiness? I am looking for a functional solution rather than an elegant one.

**Comment by tinkerborg** [comment]
@darnells Check out my pull request above, let me know if you run into any trouble.

**Comment by darnells** [comment]
@tinkerborg I finally have time to focus on this again. This is what I am seeing with your pull request:

```
==> xenserver-iso: Step: Create Instance
==> xenserver-iso: Created instance '4a68b6cd-3e40-3dff-6080-60aaba3f186c'
==> xenserver-iso: Step: Start VM Paused
==> xenserver-iso: Step: forward the instances VNC port over SSH
==> xenserver-iso: Unable to get VNC port (is the VM running?): ssh: handshake failed: ssh: unable to authenticate, attempted methods [password none], no supported methods remain
==> xenserver-iso: Destroying VM
==> xenserver-iso: Destroying VDI
```

I am using XenServer 7, packer 0.12.3, and go 1.8 using a plug-in built from xenserver/packer-builder-xenserver master/head with fetch/checkout tinkerborg/packer-builder-xenserver fix-build-ssh. The json build definition I am using is modified from the ```examples``` directory to use CentOS 6.8:

```
{
  "builders": [
    {
      "type": "xenserver-iso",
      "remote_host": "MY_IP_ADDRESS",
      "remote_username": "MY_USERNAME",
      "remote_password": "MY_PASSWROD",

      "boot_command": [
        "<tab> text ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/centos6-ks.cfg<enter><wait>"
      ],
      "boot_wait": "10s",
      "disk_size": 40960,
      "http_directory": "http",
      "iso_checksum": "28cd663c2267676414496f0929ce7bb285bf2506",
      "iso_checksum_type": "sha1",
      "iso_url": "{{user `mirror`}}/6.8/isos/x86_64/CentOS-6.8-x86_64-minimal.iso",
      "tools_iso_name": "guest-tools.iso",
      "vm_other_config": {
          "conversionvm":"true"
      },
      "network_names": [
          "MY_NETWORK_NAME"
      ],
      "output_directory": "packer-centos-6.8-x86_64-xenserver",
      "shutdown_command": "/sbin/halt",
      "ssh_username": "root",
      "ssh_password": "vmpassword",
      "ssh_wait_timeout": "10000s",
      "vm_name": "packer-centos-6.8-x86_64",
      "vm_description": "Build time: {{isotime}}"
    }
  ],

  "variables": {
    "mirror": "http://www.mirrorservice.org/sites/mirror.centos.org"
  }
}
```

**Comment by darnells** [comment]
And here is the stderr with env PACKER_LOG=1:

```
2017/03/21 22:39:22 ui: ==> xenserver-iso: Step: forward the instances VNC port over SSH
==> xenserver-iso: Step: forward the instances VNC port over SSH
2017/03/21 22:39:24 ui error: ==> xenserver-iso: Unable to get VNC port (is the VM running?): ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
==> xenserver-iso: Unable to get VNC port (is the VM running?): ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
2017/03/21 22:39:25 packer-builder-xenserver-iso: time="2017-03-21T22:39:25-05:00" level=error msg="Encountered an API error: Failure [VBD_NOT_UNPLUGGABLE OpaqueRef:a4faf485-7b64-afbb-89ff-87b4e65d1732]"
2017/03/21 22:39:25 packer-builder-xenserver-iso: 2017/03/21 22:39:25 Detached VDI 'a54dee65-fa71-4ace-bad1-edbd53c00171'
2017/03/21 22:39:25 packer-builder-xenserver-iso: time="2017-03-21T22:39:25-05:00" level=error msg="Encountered an API error: Failure [VBD_NOT_UNPLUGGABLE OpaqueRef:af904539-4d1a-3f82-2633-46e1aeca6ff5]"
2017/03/21 22:39:25 packer-builder-xenserver-iso: 2017/03/21 22:39:25 Detached VDI 'edb1cd82-4f2b-497e-bc36-d973a570864e'
2017/03/21 22:39:25 ui: ==> xenserver-iso: Destroying VM
==> xenserver-iso: Destroying VM
2017/03/21 22:39:25 packer-builder-xenserver-iso: time="2017-03-21T22:39:25-05:00" level=error msg="Encountered an API error: Failure [VM_BAD_POWER_STATE OpaqueRef:092ef771-c2b5-c7bd-d4ce-b5349d553804 paused, suspended, running halted]"
2017/03/21 22:39:25 ui: ==> xenserver-iso: Destroying VDI
```

**Comment by darnells** [comment]
@rdobson In my last comments, I reported that I was still seeing build failures when using the PR from @tinkerborg. I recommend reopening this issue. However, I will be happy to retest using the merged PR in xenserver/packer-builder-xenserver master/head just in case my previous merge attempt was invalid.

**Comment by tinkerborg** [comment]
@darnells Sorry, been a hectic time at work lately so I didn't notice your comment til I saw the PR accepted notification. Do all servers in your pool have the same root password? If not then my fix won't work. Maybe a better fix for this would to use `StartOn(host *Host, paused, force bool)` to ensure  the packer VM starts on `remote_host`. That would mean that the build would fail if the host you specified has insufficient resources to start the VM, even if another host in the pool has resources available.

**Comment by darnells** [comment]
@tinkerborg Yes, all servers in my pool have the same root password. More specifically, my pool has only one server (setup by my IT). When you mentioned root password, it occurred to me that my prior attempts were performed using my Pool Operator user. I switched credentials to the root user and was able to proceed much further beyond the `Connecting to the VM over VNC` step. Before, the VM would be destroyed within seconds after starting.

The VM is up but the XenServer Tools are not installed, thus the IP address is not reported in XenCenter Networking, and the build process stalls at `==> xenserver-iso: Waiting for SSH to become available...`. I can confirm that the SSH service is running by fetching the IP address from the VM by accessing it using XenCenter Console and then manually logging in from an SSH client. Upon closer inspection, the IP address I used to login by SSH is not the same as the one referenced in the xenserver-iso stdout (xenserver-iso: Got IP 'xx.xx.xx.xx' from HTTP request`).

Just out of curiosity, I also manually installed XenServer Tools on the VM and rebooted. XenCenter is reporting the tools are installed and the IP address is known in the Networking tab. BTW, it is still the same IP address I manually retrieved, not the IP xenserver-iso is referencing.

I have three questions:

1. What user role is required to use packer-builder-xenserver?
2. To install XenServer Tools, do I need to do something else other than define `"tools_iso_name": "guest-tools.iso"` (I am using XenServer 7.0)?
3. Are these questions and the IP mismatch better served in a new (or possibly existing) issue?
